### ISSUE TITLE
HTTP2: Use/Create new connections when maximum active streams is reached

### ISSUE THREAD

**Issue by siavashta** [reactor/reactor-netty#1789]
the netty reactor does not use/create a new connection when the max concurrent stream is reached.

gitter discussion: https://gitter.im/reactor/reactor-netty?at=6126f8df5cfd665e52edf858

Netty throws `Max active streams is reached.; nested exception is java.io.IOException: Error while acquiring from reactor.netty.internal.shaded.reactor.pool.SimpleDequePool@8ae0ed. Max active streams is reached.` while the max connection is not reached.

## Expected Behavior
a new connection to be used.

## Actual Behavior
Netty throws `Max active streams is reached.

## Steps to Reproduce
I added the client/server code here: https://github.com/siavashta/sia-h2

The max connection on the client-side is set to 100
```java
    @Bean
    public HttpClient h2Client() {

        var provider = ConnectionProvider.builder("sia")
                .maxConnections(100)
                .build();

        return HttpClient.create(provider)
                .protocol(HttpProtocol.H2C)
                .wiretap(true);
    }
```
the max concurrent stream for the server is set to `1`
```java
		DisposableServer server =
				HttpServer.create()
						.protocol(HttpProtocol.H2C)
						.http2Settings(h2 -> h2.maxConcurrentStreams(1))
						.port(9192)
						.handle((req, res) -> res.sendString(Mono
								.just("sia")
								.delayElement(Duration.ofSeconds(5))))
						.wiretap(true)
						.bindNow();

		server.onDispose()
				.block();
```
running server and client
```
stahmureszadeh@TOR-MBP-7724 siaserver % mvn spring-boot:run
stahmureszadeh@TOR-MBP-7724 siaclient % mvn spring-boot:run
```
open jmeter and start the thread group with 5 concurrent requests to client
```
stahmureszadeh@TOR-MBP-7724 sia-h2 % jmeter -t siaclient/client.jmx
```

clients encountered `Max active streams is reached.` exception
```
org.springframework.web.reactive.function.client.WebClientRequestException: Error while acquiring from reactor.netty.internal.shaded.reactor.pool.SimpleDequePool@8ae0ed. Max active streams is reached.; nested exception is java.io.IOException: Error while acquiring from reactor.netty.internal.shaded.reactor.pool.SimpleDequePool@8ae0ed. Max active streams is reached.
```

## Possible Solution
1- Create the max connections in initialization and use them properly
2- retry better on creating a new connection

**Comment by stahmureszadeh-r7** [comment]
Copying the possible solution (Least stream used strategy) from @violetagg on gitter discussion
> Violeta Georgieva
the first acquire will choose the connection with least number of streams, if all connections in the http2 pool are full then we will ask the http1 pool for a connection (regardless whether it is an idle connection or the pool will create a new one)

**Comment by crankydillo** [comment]
Hey Violeta,

I know you must like getting 'when will this be done' questions on an open source project, but.. Our current thinking is addressing this will be pretty critical to our needs.   We're still sorting that out, but we're also wondering where this lands in your work plans.  I know you have added it to the 1.0.x backlog.  Is that backlog sorted in the order you think they should be worked?  It is possible we could work on this problem, but the code is moderately dense in this area;)

**Comment by violetagg** [comment]
yeah ... I'll try to 1.0.11 but cannot promise ... most probably 1.0.12

**Comment by crankydillo** [comment]
No promises required!   Just getting a sense of your plans/timelines:)

**Comment by philip-jvm** [comment]
We are also experiencing this. 
As a workaround for thousands of concurrent streams we are using HTTP11 with a large ConnectionProvider#maxConnections
Thanks for all the work on this project!

**Comment by violetagg** [comment]
A change about this issue is available in 1.0.14-SNAPSHOT. It will be great if you are able to test it.

**Comment by rtonini-r7** [comment]
@violetagg We have tested this change locally and no longer get `Max active streams is reached`.  Thanks!
